File: 1250-longest-common-subsequence/longest-common-subsequence.py
# class Solution:
#     def longestCommonSubsequence(self, s1: str, s2: str) -> int:
#         dp=[[0 for j in range(len(s2)+1)] for i in range(len(s1)+1)]

#         for i in range(len(s1)-1,-1,-1):
#             for j in range(len(s2)-1,-1,-1):
#                 if(s1[i]==s2[j]):
#                     dp[i][j]=1+dp[i+1][j+1]
#                 else:
#                     dp[i][j]=max(dp[i+1][j],dp[i][j+1])
#         return dp[0][0]

class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        dp = [[0 for i in range(len(text2) + 1)] for j in range(len(text1) + 1)]
        
        text1 = " " + text1
        text2 = " " + text2
        
        for i in range(1, len(text1)):
            for j in range(1, len(text2)):
                if text1[i] == text2[j]:
                    dp[i][j] = 1 + dp[i-1][j-1]
                else:
                    dp[i][j] = max(dp[i-1][j], dp[i][j-1])
            
        return dp[-1][-1]

File: 1250-longest-common-subsequence/longest-common-subsequence.cpp
// class Solution {
// public:
//     int longestCommonSubsequence(string text1, string text2) {
//         int m = text1.size(), n = text2.size();
//         int dp[m+1][n+1];
//         memset(dp, 0, sizeof(dp));
//         for(int i=1; i<=m; ++i) {
//             for(int j=1; j<=n; ++j) {
//                 if(text1[i-1] == text2[j-1]) dp[i][j] = 1+dp[i-1][j-1];
//                 else dp[i][j] = max(dp[i-1][j], dp[i][j-1]);
//             }
//         }
//         return dp[m][n];
//     }
// };
#include <bits/stdc++.h>
using namespace std;
class Solution {
public:
    int longestCommonSubsequence(string text1, string text2) {
        int m = text1.size(), n = text2.size();
        vector<vector<int>> dp(m+1,vector<int>(n+1));
        for(int row=0;row<m+1;row++){
            dp[row][0]=0;
        }
        for(int col=0;col<n+1;col++){
            dp[0][col]=0;
        }
        for(int i=1;i<m+1;i++){
            for(int j=1;j<n+1;j++){
                if(text1[i-1]==text2[j-1]){
                    dp[i][j]=1+dp[i-1][j-1];
                }else{
                    dp[i][j]=max(dp[i-1][j],dp[i][j-1]);
                }
            }
        }
        return dp[m][n];
    }
};